File: core/src/com/jupiter/europa/entity/ability/Target.java
package com.jupiter.europa.entity.ability;

import com.badlogic.ashley.core.Entity;

import java.awt.*;

/**
 * Created by nathan on 5/21/15.
 */
public abstract class Target {

    public static TileTarget tile(int x, int y) {
        return tile(new Point(x, y));
    }

    public static TileTarget tile(Point point) {
        return new TileTarget(point);
    }

    public static EntityTarget entity(Entity entity) {
        return new EntityTarget(entity);
    }


    public static class TileTarget extends Target {

        // Fields
        public final Point tile;


        // Initialization
        public TileTarget(Point point) {
            this.tile = point;
        }
    }

    public static class EntityTarget extends Target {

        // Fields
        public final Entity entity;


        // Initialization
        private EntityTarget(Entity entity) {
            this.entity = entity;
        }
    }
}

File: core/src/com/jupiter/europa/entity/ability/Target.kt
/*
 * The MIT License
 *
 * Copyright 2015 Nathan Templon.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

package com.jupiter.europa.entity.ability

import com.badlogic.ashley.core.Entity
import java.awt.Point

/**
 * Created by nathan on 5/21/15.
 */
public abstract class Target {


    public class TileTarget(public val tile: Point) : Target()

    public class EntityTarget internal constructor(public val entity: Entity) : Target()

    companion object {
        public fun tile(x: Int, y: Int): TileTarget {
            return tile(Point(x, y))
        }

        public fun tile(point: Point): TileTarget {
            return TileTarget(point)
        }

        public fun entity(entity: Entity): EntityTarget {
            return EntityTarget(entity)
        }
    }
}
